curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java

```diff
@@ -240,11 +238,7 @@ public void createContainers(String path) throws Exception
     @Override
     public void clearWatcherReferences(Watcher watcher)
     {
-        NamespaceWatcher namespaceWatcher = namespaceWatcherMap.remove(watcher);
-        if ( namespaceWatcher != null )
-        {
-            namespaceWatcher.close();
-        }
```

Comment:
Recipes have to do a tiny bit of work. They need to get a specially wrapped instance by calling CuratorFramework.newWatcherRemoveCuratorFramework() and then calling removeWatchers() when done.